### DISCUSSION TITLE
Adding tabs

### DISCUSSION THREAD

**Discussion by zprofreelancer** [serenity-is/Serenity#5388] [Help] [answered]
hello,
i'm trying to add tabs in a page. so, i saw the "entity dialog as panel" page that contain tabs but i can't figure out how is the tabs done..Also i want to edit the tabs title. i have read the tutorials... the tabbed template example differs from implementation in the "entity dialog as panel" example..i need an explanation of how the tabs is done in the "entity dialog as panel" example . So, could you please help me with that?.

Thanks

**Comment by minhhungit** [comment, marked as answer]
`EntityDialogAsPanel` is extended from [Northwind.OrderDialog](https://github.com/serenity-is/Serene/blob/8f0f778d1331d3988610854e0a85e14dbff20757/Serene/Serene.Core/Modules/Northwind/Order/OrderDialog.ts#L5)

`Northwind.OrderDialog` uses [OrderForm](https://github.com/serenity-is/Serene/blob/8f0f778d1331d3988610854e0a85e14dbff20757/Serene/Serene.Core/Modules/Northwind/Order/OrderForm.cs)

You will see tab is "defined" in `OrderForm` 


![image](https://user-images.githubusercontent.com/2279508/105648912-7eebcd00-5ee0-11eb-9afe-16cc3c98f67d.png)

**Reply by zprofreelancer** [reply]
Thanks alot...i couldn't reach the orderform.cs because when i press f12 on orderform used in orderdialog, it goes to "northwind.orderform.ts".
why the "go to definition" does this?...is there a solution for that?

Thanks.

**Comment by minhhungit** [comment]
![image](https://user-images.githubusercontent.com/2279508/105648945-a93d8a80-5ee0-11eb-8445-829542d96505.png)

**Reply by zprofreelancer** [reply]
thanks...i don't know why it doesn't appear at the definitions when i press F12 or click go to definitions...

**Comment by minhhungit** [comment]
`northwind.orderform.ts` is generated from `OrderForm.cs` by Serenity Code Generator 
If you are using MVC version, then it is generated when you transform t4 template, if you are using .net core version, you just need build your project

**Reply by zprofreelancer** [reply]
Thanks...I have installed serene mvc template and build the project and run it... I works fine....Should i choose build --> transfer all templates or since i already have the ".ts" files then the template was automatically transformed??

**Reply by zprofreelancer** [reply]
> 
> 
> `northwind.orderform.ts` is generated from `OrderForm.cs` by Serenity Code Generator
> If you are using MVC version, then it is generated when you transform t4 template, if you are using .net core version, you just need build your project

ok...But why the .ts generated file only appear after pressing f12 and the .cs doesn't appear?

**Reply by minhhungit** [reply]
OrderForm.cs is what you will create, or it also can be generated when you use `sergen` tool to generate code from your table

After you have OrderForm.cs, transform t4 template will generate typescript class for client side

what is F12, I don't know, I did not say anything about it

And again, please read document and do tutorial, Volkan explain very detail in the document.

If my anwser help you know how to add new tab to your form (dialog), please mark it as anwsered so other people will know that it is solution

**Reply by zprofreelancer** [reply]
> OrderForm.cs is what you will create, or it also can be generated when you use `sergen` tool to generate code from your table
> 
> After you have OrderForm.cs, transform t4 template will generate typescript class for client side
> 
> what is F12, I don't know, I did not say anything about it
> 
> And again, please read document and do tutorial, Volkan explain very detail in the document.
> 
> If my anwser help you know how to add new tab to your form (dialog), please mark it as anwsered so other people will know that it is solution

Thanks alot....i'm already reading the tutorial but i haven't found any mention about .cs tabbed form..Sorry, i forgot to mark the post...now, i marked it... again, thanks alot for your help.
